RecoTracker/MkFit/plugins/MkFitProducer.cc

```diff
@@ -0,0 +1,148 @@
+#include "mkFit/MkBuilderWrapper.h"
+
+// TBB includes
+#include "tbb/task_arena.h"
+
+// std includes
+#include <functional>
+
+class MkFitProducer : public edm::global::EDProducer<edm::StreamCache<mkfit::MkBuilderWrapper> > {
+public:
+  explicit MkFitProducer(edm::ParameterSet const& iConfig);
+  ~MkFitProducer() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  std::unique_ptr<mkfit::MkBuilderWrapper> beginStream(edm::StreamID) const override;
+
+private:
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+
+  edm::EDGetTokenT<MkFitInputWrapper> hitsSeedsToken_;
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomToken_;
+  edm::EDPutTokenT<MkFitOutputWrapper> putToken_;
+  std::function<double(mkfit::Event&, mkfit::MkBuilder&)> buildFunction_;
+  bool backwardFitInCMSSW_;
+  bool mkfitSilent_;
+};
+
+MkFitProducer::MkFitProducer(edm::ParameterSet const& iConfig)
+    : hitsSeedsToken_{consumes<MkFitInputWrapper>(iConfig.getParameter<edm::InputTag>("hitsSeeds"))},
+      geomToken_{esConsumes<TrackerGeometry, TrackerDigiGeometryRecord>()},
+      putToken_{produces<MkFitOutputWrapper>()},
+      backwardFitInCMSSW_{iConfig.getParameter<bool>("backwardFitInCMSSW")},
+      mkfitSilent_{iConfig.getUntrackedParameter<bool>("mkfitSilent")} {
+  const auto build = iConfig.getParameter<std::string>("buildingRoutine");
+  bool isFV = false;
+  if (build == "bestHit") {
+    buildFunction_ = mkfit::runBuildingTestPlexBestHit;
+  } else if (build == "standard") {
+    buildFunction_ = mkfit::runBuildingTestPlexStandard;
+  } else if (build == "cloneEngine") {
+    buildFunction_ = mkfit::runBuildingTestPlexCloneEngine;
+  } else if (build == "fullVector") {
+    isFV = true;
+    buildFunction_ = mkfit::runBuildingTestPlexFV;
+  } else {
+    throw cms::Exception("Configuration") << "Invalid value for parameter 'buildingRoutine' " << build
+                                          << ", allowed are bestHit, standard, cloneEngine, fullVector";
+  }
+
+  const auto seedClean = iConfig.getParameter<std::string>("seedCleaning");
+  auto seedCleanOpt = mkfit::ConfigWrapper::SeedCleaningOpts::noCleaning;
+  if (seedClean == "none") {
+    seedCleanOpt = mkfit::ConfigWrapper::SeedCleaningOpts::noCleaning;
+  } else if (seedClean == "N2") {
+    seedCleanOpt = mkfit::ConfigWrapper::SeedCleaningOpts::cleanSeedsN2;
+  } else {
+    throw cms::Exception("Configuration")
+        << "Invalida value for parameter 'seedCleaning' " << seedClean << ", allowed are none, N2";
+  }
+
+  auto backwardFitOpt =
+      backwardFitInCMSSW_ ? mkfit::ConfigWrapper::BackwardFit::noFit : mkfit::ConfigWrapper::BackwardFit::toFirstLayer;
+
+  // TODO: what to do when we have multiple instances of MkFitProducer in a job?
+  mkfit::MkBuilderWrapper::populate(isFV);
+  mkfit::ConfigWrapper::initializeForCMSSW(seedCleanOpt, backwardFitOpt, mkfitSilent_);
```

Comment:
What number of threads? The number of TBB threads is set by the framework.

Reply:
I meant the number of threads per module call.

Reply:
I'm confused. Tasks spawned by a thread (by the `tbb::parallel_for`s) may be run by other threads if those threads run out of their own work. I'm not aware of any way to tell TBB to run a `parallel_for` in a given number of threads.

(there is a way to set a maximum number of threads for a task arena, but I suppose you're not after that)

Reply:
mkfit can run multiple threads per event.
the default configuration is apparently 1 thread.
To be able to do some scale tests I was asking to add a configurable parameter for the parallel_for etc in mkFit to be able to use more threads.

I didn't think of a specific implementation when asking.

some specific examples I guess are in https://github.com/cms-sw/cmssw/blob/74becf9d33ac741b0e4d92b292b1d6cf61619d47/RecoLocalCalo/HGCalRecProducers/plugins/HGCalImagingAlgo.cc#L96-L98
but you likely know better which can fit best in this case.

Reply:
> the default configuration is apparently 1 thread.

Default configuration of what?

> To be able to do some scale tests I was asking to add a configurable parameter for the parallel_for etc in mkFit to be able to use more threads.

I'm not aware of any parameter that could be added.

AFAICT the best that can be done to test the parallelism inside mkFit is to run the CMSSW job with one EDM stream and variable number of threads, and make sure that there are no other modules with internal parallelism.

> some specific examples I guess are in
> 
> https://github.com/cms-sw/cmssw/blob/74becf9d33ac741b0e4d92b292b1d6cf61619d47/RecoLocalCalo/HGCalRecProducers/plugins/HGCalImagingAlgo.cc#L96-L98
> 
> 
> but you likely know better which can fit best in this case.

The `parallel_for` arguments do not set the number of threads. TBB internally breaks the `parallel_for` iterations into tasks, and runs these tasks as it wishes (that can be "all tasks in one thread", "all tasks in different threads", or anything in between).

Reply:
I may have also misread the config parameters in mkfit and we already have the ability to run multi-thread calls depending on the framework mix of stream/thread settings.
Please clarify the current state.

Reply:
> TBB internally breaks the `parallel_for` iterations into tasks

Let me add that there is a way to control this chunking of the work of the loop into tasks, but that does not affect the TBB task scheduler.

Reply:
> I may have also misread the config parameters in mkfit and we already have the ability to run multi-thread calls depending on the framework mix of stream/thread settings.
> Please clarify the current state.

Everything from the CMSSW side has been constructed to support calls from multiple threads with any mix of streams/threads.

The only known multi-threading issue is that having more than one instance of `MkFitProducer` would lead to data races in initialization of mkFit.